Add a function for binary exponentiation

samarth491 opened the pull request:
A function for binary exponentiation is added which can calculate
x raised to power y modulo p in O(log(y)) time. This can be very
helpful when y is large.

ref. #6

r1walz commented on Fundamentals/bin_expo.cpp:
```diff
@@ -0,0 +1,52 @@
+
+using namespace std;
+
+
```
Remove this line.

r1walz commented on Fundamentals/bin_expo.cpp:
```diff
@@ -0,0 +1,52 @@
+ * Time complexity: O(log(y))
+ * Space Complexity: O(1)
+ */
+
```
Remove this line too.

r1walz commented on Fundamentals/bin_expo.cpp:
```diff
@@ -0,0 +1,52 @@
+
+	x %= p;
+
+	while(y) {
```
Space after while, follow coding guidelines. See CONTRIBUTING.md.

r1walz commented on Fundamentals/bin_expo.cpp:
```diff
@@ -0,0 +1,52 @@
+	x %= p;
+
+	while(y) {
+		if(y%2)
```
Follow coding guidelines.

r1walz commented on Fundamentals/bin_expo.cpp:
```diff
@@ -0,0 +1,52 @@
+}
+
+int main() {
+
```
Remove this line.

r1walz commented on Fundamentals/bin_expo.cpp:
```diff
@@ -0,0 +1,52 @@
+
+	int x_raised_y = bin_expo(x, y, p);
+
+	cout<<x_raised_y<<endl;
```
Add spaces.

r1walz commented on Fundamentals/bin_expo.cpp:
```diff
@@ -0,0 +1,52 @@
+ * Space Complexity: O(1)
+ */
+
+int bin_expo(int x, int y, int p) {
```
Can you also add recursive implementation too?

r1walz commented on Fundamentals/bin_expo.cpp:
```diff
@@ -0,0 +1,52 @@
+
+	while(y) {
+		if(y%2)
+			res = (res * 1LL * x) % p;
```
Type cast res instead of multiplying by 1LL, it saves one instruction.

```cpp
res = ((long long)res * x) % p;
```

and justify why is it used as it may not be apparent to beginners.

r1walz reviewed (commented):
This looks good. Share your views and continue. Thanks!

samarth491 committed:
Add function for binary exponentiation
A function for binary exponentiation is added which can calculate
x raised to power y modulo p in O(log(y)) time. This can be very
helpful when y is large.
ref. #6
